pkg/codegen/java/gen.go

```diff
@@ -1515,50 +1534,101 @@ func (mod *modContext) genFunctions(ctx *classFileContext, addClass addClassMeth
 			fprintf(w,
 				"        return %s(%s.Empty, %s.Empty);\n",
 				methodName, argsType, invokeOptions)
 			fprintf(w, "    }\n")
 
+			// Generate invoke that return CompletableFuture<T>
+			printCommentFunction(ctx, fun, indent)
+			fprintf(w, "    public static %s<%s> %s() {\n",
+				ctx.ref(names.CompletableFuture), returnType, plainMethodName)
+			fprintf(w,
+				"        return %s(%s.Empty, %s.Empty);\n",
+				plainMethodName, plainArgsType, invokeOptions)
+			fprintf(w, "    }\n")
 		}
 
-		// Add args only invoke
+		// Output version: add args only invoke
 		printCommentFunction(ctx, fun, indent)
 		fprintf(w, "    public static %s<%s> %s(%s args) {\n",
-			ctx.ref(names.CompletableFuture), returnType, methodName, argsType)
+			ctx.ref(names.Output), returnType, methodName, argsType)
 		fprintf(w,
 			"        return %s(args, %s.Empty);\n",
 			methodName, invokeOptions)
 		fprintf(w, "    }\n")
 
-		// Add full invoke
+		// CompletableFuture version: add args only invoke
+		printCommentFunction(ctx, fun, indent)
+		fprintf(w, "    public static %s<%s> %s(%s args) {\n",
+			ctx.ref(names.CompletableFuture), returnType, plainMethodName, plainArgsType)
+		fprintf(w,
+			"        return %s(args, %s.Empty);\n",
+			plainMethodName, invokeOptions)
+		fprintf(w, "    }\n")
+
+		// Output version: add full invoke
+		printCommentFunction(ctx, fun, indent)
+		fprintf(w, "    public static %s<%s> %s(%s args, %s options) {\n",
+			ctx.ref(names.Output), returnType, methodName, argsType, invokeOptions)
+		fprintf(w,
+			"        return %s.getInstance().invoke(\"%s\", %s.of(%s.class), args, %s.withVersion(options));\n",
+			ctx.ref(names.Deployment), fun.Token, ctx.ref(names.TypeShape), returnType, mod.utilitiesRef(ctx))
+		fprintf(w, "    }\n")
+
+		// CompletableFuture version: add full invoke
+		// notice how the implementation now uses `invokeAsync` instead of `invoke`
 		printCommentFunction(ctx, fun, indent)
 		fprintf(w, "    public static %s<%s> %s(%s args, %s options) {\n",
-			ctx.ref(names.CompletableFuture), returnType, methodName, argsType, invokeOptions)
+			ctx.ref(names.CompletableFuture), returnType, plainMethodName, plainArgsType, invokeOptions)
 		fprintf(w,
 			"        return %s.getInstance().invokeAsync(\"%s\", %s.of(%s.class), args, %s.withVersion(options));\n",
 			ctx.ref(names.Deployment), fun.Token, ctx.ref(names.TypeShape), returnType, mod.utilitiesRef(ctx))
 		fprintf(w, "    }\n")
 
 		// Emit the args and result types, if any.
 		if fun.Inputs != nil {
+			// Generate "{Function}Args" class for invokes that return Output<T>
 			if err := addClass(inputsPkg, argsClass, func(ctx *classFileContext) error {
 				args := &plainType{
 					mod:                   mod,
 					name:                  ctx.className.String(),
 					baseClass:             "com.pulumi.resources.InvokeArgs",
 					propertyTypeQualifier: inputsQualifier,
 					properties:            fun.Inputs.Properties,
+					propertiesAsOutputs:   true,
```

Comment:
This doesn't really work I'm afraid think for nested types.. the proeprties

```
f(FArgs)
class FArgs {
  A Output<AType>;
}
class AType {
  B Output<Integer>; // want this to vary also, `Integer` vs `Output<Integer>`
}
```

Reply:
I am wondering if the InputShape facility can just work for us here. Like this:

https://github.com/pulumi/pulumi/blob/ecf65dab2055f1c187d03a2d90aa200167785f76/pkg/codegen/dotnet/gen.go#L1474

You use `fun.Inputs.InputShape` and that automagically does the underlying conversion of properties to Output-wrapped properties deeply - that's my hope at least. Could we explore using that and removing `propertiesAsOutputs:   true,` and manual attempts to wrap in `Output` here?